### PR TITLE
Take account of `Entity` model in `BracketRequestResponse` Middleware

### PR DESCRIPTION
**Pull request by danicheg**
... for better performance.

### CODE REVIEW

File: server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala
Comment (danicheg) [inline server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala:150]: _Nota bene_: since the response body has been materialized for `Empty`/`Strict` Entities, we are running the release eagerly. And that obviously differs from the behavior of streamed `Entity`, for which release will be logically delayed until `Stream` is materialized. Hence why we have to prevent attempting release twice for non-streamed Entities.

File: server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala
Comment (isomarcte) [inline server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala:301]: Can we extend `RuntimeException` or `Exception` instead?

Comment (isomarcte) [inline server/shared/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala:301]: Can you change `cause` to `override val getCause: Throwable`?

**Comment by danicheg** [comment]
@isomarcte your suggestions make sense to me, and they have been addressed. Thanks for reviewing, by the way!

**Review by isomarcte — APPROVED** [review]
LGTM!

**Comment by danicheg** [comment]
I'm happy to merge this as minimal requirements are successfully passed, but feel yet another pair of eyes is needed to glance at this before moving on. If nobody minds, I will wait for one more week before merging this.

**Pull request merged**
